assert_type() not matching function type

**Bug Report**

'assert_type()' refuses to match a function to its own type.

**To Reproduce**

```python
from typing import assert_type, Callable
def myfunc(arg: int) -> None: pass
assert_type(myfunc, Callable[[int], None])

# Gist URL: https://gist.github.com/mypy-play/50dde8e891d971d823c4c6af2e639834
# Playground URL: https://mypy-play.net/?mypy=latest&python=3.11&gist=50dde8e891d971d823c4c6af2e639834
```

**Expected Behavior**

No error.

**Actual Behavior**

```
main.py:5: error: Expression is of type "Callable[[int], None]", not "Callable[[int], None]"  [assert-type]
Found 1 error in 1 file (checked 1 source file)
```

The error disappears if the function takes no arguments.

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: both 1.2.0 and current master branch (commit [29cc561](https://github.com/python/mypy/commit/29cc5614b94d2357794fe5c62c64525fd8087e59))
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): default
- Python version used: 3.11

This has been fixed on the master branch. The error message now shows: 

```
main.py:5: error: Expression is of type "Callable[[Arg(int, 'arg')], None]", not "Callable[[int], None]"  [assert-type]
Found 1 error in 1 file (checked 1 source file)
```

You can see this in the playground: https://mypy-play.net/?mypy=master&python=3.11&gist=50dde8e891d971d823c4c6af2e639834

--- a/test-data/unit/check-assert-type-fail.test
+++ b/test-data/unit/check-assert-type-fail.test
@@ -26,3 +26,8 @@ class array:
 def f(si: arr.array[int]):
     typing.assert_type(si, int) # E: Expression is of type "array[int]", not "int"
 [builtins fixtures/tuple.pyi]
+
+[case testAssertTypeFailCallableArgKind]
+from typing import assert_type, Callable
+def myfunc(arg: int) -> None: pass
+assert_type(myfunc, Callable[[int], None])  # E: Expression is of type "Callable[[Arg(int, 'arg')], None]", not "Callable[[int], None]"

